// aop-ui/lib/_components/@alife/next-form/lib/submit.js
'use strict';

exports.__esModule = true;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _class, _temp2;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _nextButton = require('../../next-button/lib/index.js');

var _nextButton2 = _interopRequireDefault(_nextButton);

var _nextUtil = require('../../next-util/lib/index.js');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Form.Submit
 * @description 继承 Button API
 */
var Submit = (_temp2 = _class = function (_React$Component) {
    (0, _inherits3.default)(Submit, _React$Component);

    function Submit() {
        var _temp, _this, _ret;

        (0, _classCallCheck3.default)(this, Submit);

        for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
            args[_key] = arguments[_key];
        }

        return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, _React$Component.call.apply(_React$Component, [this].concat(args))), _this), _this.handleClick = function () {
            var field = _this.context._formField || _this.props.field;
            var validate = _this.props.validate;
            if (validate === true) {
                field.validate(function (errors) {
                    _this.props.onClick(field.getValues(), errors);
                });
            } else if (Array.isArray(validate)) {
                field.validate(validate, function (errors) {
                    _this.props.onClick(field.getValues(), errors);
                });
            } else {
                _this.props.onClick(field.getValues());
            }
        }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
    }

    Submit.prototype.render = function render() {
        var children = this.props.children;


        return _react2.default.createElement(
            _nextButton2.default,
            (0, _extends3.default)({}, _nextUtil.obj.pickOthers(Submit.propTypes, this.props), { onClick: this.handleClick }),
            children
        );
    };

    return Submit;
}(_react2.default.Component), _class.propTypes = {
    /**
     * 点击提交后触发
     */
    onClick: _propTypes2.default.func,
    /**
     * 是否校验/需要校验的 name 数组
     */
    validate: _propTypes2.default.oneOfType([_propTypes2.default.bool, _propTypes2.default.array]),
    /**
     * 自定义 field (在 Form 内不需要设置)
     */
    field: _propTypes2.default.object,
    children: _propTypes2.default.node
}, _class.defaultProps = {
    onClick: _nextUtil.func.noop
}, _class.contextTypes = {
    _formField: _propTypes2.default.object
}, _class._typeMark = 'form_item', _temp2);
Submit.displayName = 'Submit';
exports.default = Submit;
module.exports = exports['default'];
