// aop-ui/lib/_components/@alife/next-locale/lib/en-us.js
'use strict';

exports.__esModule = true;
exports.default = {
    momentLocale: 'en',
    Timeline: {
        expand: 'Expand',
        fold: 'Fold'
    },
    Card: {
        expand: 'Expand',
        fold: 'Fold'
    },
    Calendar: {
        today: 'Today',
        now: 'Now',
        ok: 'Ok',
        clear: 'Clear',
        month: 'Month',
        year: 'Year',
        prevYear: 'Previous year',
        nextYear: 'Next year',
        prevMonth: 'Previous month',
        nextMonth: 'Next month',
        prevDecade: 'Previous decade',
        nextDecade: 'Next decade'
    },
    DatePicker: {
        placeholder: 'Select date',
        datetimePlaceholder: 'Select date and time',
        now: 'Now',
        selectTime: 'Select time',
        selectDate: 'Select date',
        ok: 'OK',
        clear: 'Clear',
        startPlaceholder: 'Start date',
        endPlaceholder: 'End date'
    },
    Dialog: {
        ok: 'Ok',
        cancel: 'Cancel'
    },
    Pagination: {
        prev: 'Previous',
        next: 'Next',
        goTo: 'Go to',
        page: 'Page',
        go: 'Go',
        pageSize: 'Items per page:'
    },
    Select: {
        selectPlaceHolder: 'Please select',
        autoCompletePlaceHolder: 'Please input'
    },
    Table: {
        empty: 'No data',
        ok: 'Ok',
        reset: 'Reset'
    },
    TimePicker: {
        placeholder: 'Select time',
        clear: 'Clear',
        hour: 'H',
        minute: 'M',
        second: 'S'
    },
    Transfer: {
        items: 'items',
        item: 'item',
        moveAll: 'Move All',
        searchPlaceholder: 'Please input'
    },
    Upload: {
        card: {
            cancel: 'Cancel',
            addPhoto: 'Add Picture'
        },
        drag: {
            text: 'Click or Drag the file to this area to upload',
            hint: 'Support docx, xls, PDF, rar, zip, PNG, JPG and other files upload'
        }
    }
};
module.exports = exports['default'];
