// aop-ui/lib/_components/@alife/next-locale/lib/ja-jp.js
'use strict';

exports.__esModule = true;
exports.default = {
    momentLocale: 'ja',
    Timeline: {
        expand: '開く',
        fold: '閉じる'
    },
    Card: {
        expand: '開く',
        fold: '閉じる'
    },
    Calendar: {
        today: '本日',
        now: '現在',
        ok: '確認',
        clear: 'クリア',
        month: '月',
        year: '年',
        prevYear: '前の年',
        nextYear: '次の年',
        prevMonth: '前の月',
        nextMonth: '来月',
        prevDecade: '前の十年',
        nextDecade: '次の十年'
    },
    DatePicker: {
        placeholder: '日付けの選択',
        datetimePlaceholder: '日時を選択してください',
        now: '現在',
        selectTime: '時間の選択',
        selectDate: '日付けの選択',
        ok: '確認',
        clear: 'クリア',
        startPlaceholder: '開始日',
        endPlaceholder: '最終日'
    },
    Dialog: {
        ok: '確認',
        cancel: 'キャンセル'
    },
    Pagination: {
        prev: '前のページ',
        next: '次のページ',
        goTo: '第',
        page: 'ページへ',
        go: '確認',
        pageSize: 'ページサイズ：'
    },
    Select: {
        selectPlaceHolder: '選択',
        autoCompletePlaceHolder: '入力'
    },
    Table: {
        empty: 'データなし',
        ok: '確認',
        reset: 'リセット'
    },
    TimePicker: {
        placeholder: '時間の選択',
        clear: 'クリア',
        hour: '時',
        minute: '分',
        second: '秒'
    },
    Transfer: {
        items: '項目',
        item: '項目',
        moveAll: '全ての移動',
        searchPlaceholder: '入力'
    },
    Upload: {
        card: {
            cancel: 'キャンセル',
            addPhoto: '画像アップロード'
        },
        drag: {
            text: '枠にドラッグしてアップロード',
            hint: 'docx, xls, PDF, rar, zip, PNG, JPG 等がサポートしています'
        }
    }
};
module.exports = exports['default'];
