// aop-ui/lib/_components/@alife/next-locale/lib/zh-tw.js
'use strict';

exports.__esModule = true;
exports.default = {
    momentLocale: 'zh-tw',
    Timeline: {
        expand: '展開',
        fold: '收起'
    },
    Card: {
        expand: '展開',
        fold: '收起'
    },
    Calendar: {
        today: '今天',
        now: '此刻',
        ok: '確定',
        clear: '清除',
        month: '月',
        year: '年',
        prevYear: '上一年',
        nextYear: '下一年',
        prevMonth: '上个月',
        nextMonth: '下个月',
        prevDecade: '上十年',
        nextDecade: '后十年'
    },
    DatePicker: {
        placeholder: '請選擇日期',
        datetimePlaceholder: '請選擇日期和時間',
        now: '此刻',
        selectTime: '選擇時間',
        selectDate: '選擇日期',
        ok: '確定',
        clear: '清除',
        startPlaceholder: '起始日期',
        endPlaceholder: '結束日期'
    },
    Dialog: {
        ok: '確認',
        cancel: '取消'
    },
    Pagination: {
        prev: '上一頁',
        next: '下一頁',
        goTo: '到第',
        page: '頁',
        go: '確定',
        pageSize: '每頁顯示：'
    },
    Select: {
        selectPlaceHolder: '請選擇',
        autoCompletePlaceHolder: '請輸入'
    },
    Table: {
        empty: '沒有數據',
        ok: '確認',
        reset: '重置'
    },
    TimePicker: {
        placeholder: '請選擇時間',
        clear: '清除',
        hour: '時',
        minute: '分',
        second: '秒'
    },
    Transfer: {
        items: '項',
        item: '項',
        moveAll: '移動全部',
        searchPlaceholder: '請輸入'
    },
    Upload: {
        card: {
            cancel: '取消',
            addPhoto: '上傳圖片'
        },
        drag: {
            text: '點擊或者拖動文件到虛線框內上傳',
            hint: '支持 docx, xls, PDF, rar, zip, PNG, JPG 等類型的文件'
        }
    }
};
module.exports = exports['default'];
