// aop-ui/lib/_components/@alife/next-upload/lib/runtime/uploader.js
'use strict';

exports.__esModule = true;
exports.default = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _nextUtil = require('../../../next-util/lib/index.js');

var _util = require('../util.js');

var _request = require('./request.js');

var _request2 = _interopRequireDefault(_request);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var noop = _nextUtil.func.noop;

var Uploader = function () {
    function Uploader(options) {
        (0, _classCallCheck3.default)(this, Uploader);

        this.options = (0, _extends3.default)({
            beforeUpload: noop,
            onProgress: noop,
            onSuccess: noop,
            onError: noop,
            data: {},
            name: 'file',
            method: 'post'
        }, options);
        this.reqs = {};
    }

    Uploader.prototype.setOptions = function setOptions(options) {
        (0, _extends3.default)(this.options, options);
    };

    Uploader.prototype.startUpload = function startUpload(files) {
        var _this = this;

        var filesArr = files.length ? Array.prototype.slice.call(files) : [files];
        filesArr.forEach(function (file) {
            file.uid = file.uid || (0, _util.uid)();
            _this.upload(file);
        });
    };

    Uploader.prototype.abort = function abort(file) {
        var reqs = this.reqs;

        if (file) {
            var _uid = file;
            if (file && file.uid) {
                _uid = file.uid;
            }
            if (reqs[_uid]) {
                reqs[_uid].abort();
                delete reqs[_uid];
            }
        } else {
            Object.keys(reqs).forEach(function (uid) {
                if (reqs[uid]) {
                    reqs[uid].abort();
                }
                delete reqs[uid];
            });
        }
    };

    Uploader.prototype.upload = function upload(file) {
        var _this2 = this;

        var beforeUpload = this.options.beforeUpload;

        var before = beforeUpload(file);

        _nextUtil.func.promiseCall(before, function () {
            _this2.post(file);
        }, function () {
            _this2.options.onError(null, null, file);
        });
    };

    Uploader.prototype.post = function post(file) {
        var _this3 = this;

        var _options = this.options,
            action = _options.action,
            name = _options.name,
            headers = _options.headers,
            timeout = _options.timeout,
            withCredentials = _options.withCredentials,
            _onProgress = _options.onProgress,
            _onSuccess = _options.onSuccess,
            _onError = _options.onError,
            method = _options.method;


        var data = this.options.data;
        if (typeof data === 'function') {
            data = data(file);
        }

        var uid = file.uid;


        this.reqs[uid] = (0, _request2.default)({
            action: action,
            filename: name,
            file: file,
            data: data,
            timeout: timeout,
            headers: headers,
            withCredentials: withCredentials,
            method: method,
            onProgress: function onProgress(e) {
                _onProgress(e, file);
            },
            onSuccess: function onSuccess(ret) {
                delete _this3.reqs[uid];
                _onSuccess(ret, file);
            },
            onError: function onError(err, ret) {
                delete _this3.reqs[uid];
                _onError(err, ret, file);
            }
        });
    };

    return Uploader;
}();

exports.default = Uploader;
module.exports = exports['default'];
