// aop-ui/lib/_components/@alife/next-table/lib/base/sort.js
'use strict';

exports.__esModule = true;
exports.default = undefined;

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _class, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _nextIcon = require('../../../next-icon/lib/index.js');

var _nextIcon2 = _interopRequireDefault(_nextIcon);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* eslint-disable react/prefer-stateless-function */
var Sort = (_temp = _class = function (_React$Component) {
    (0, _inherits3.default)(Sort, _React$Component);

    function Sort() {
        (0, _classCallCheck3.default)(this, Sort);
        return (0, _possibleConstructorReturn3.default)(this, _React$Component.apply(this, arguments));
    }

    // 渲染排序
    Sort.prototype.renderSort = function renderSort() {
        var _props = this.props,
            prefix = _props.prefix,
            sort = _props.sort,
            dataIndex = _props.dataIndex,
            sortStatus = sort[dataIndex],
            map = {
            desc: 'descending',
            asc: 'ascending'
        };

        var icons = ['asc', 'desc'].map(function (sortOrder) {
            return _react2.default.createElement(
                'a',
                { href: 'javascript:;',
                    key: sortOrder,
                    className: sortStatus === sortOrder ? 'current' : '' },
                _react2.default.createElement(_nextIcon2.default, { type: map[sortOrder], size: 'small' })
            );
        });

        return _react2.default.createElement(
            'span',
            { className: prefix + 'table-sort',
                onClick: this.onSort.bind(this, dataIndex, sortStatus === 'asc' ? 'desc' : 'asc') },
            icons
        );
    };

    Sort.prototype.onSort = function onSort(dataIndex, order) {
        var sort = {};
        sort[dataIndex] = order;
        this.props.onSort(dataIndex, order, sort);
    };

    Sort.prototype.render = function render() {
        return this.renderSort();
    };

    return Sort;
}(_react2.default.Component), _class.propTypes = {
    prefix: _propTypes2.default.string,
    className: _propTypes2.default.string,
    sort: _propTypes2.default.object,
    onSort: _propTypes2.default.func,
    dataIndex: _propTypes2.default.string
}, _class.defaultProps = {
    sort: {} }, _temp);
Sort.displayName = 'Sort';
exports.default = Sort;
module.exports = exports['default'];
