// aop-ui/lib/_components/@alife/next-step/lib/view/step-item.js
'use strict';

exports.__esModule = true;

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _class, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _nextIcon = require('../../../next-icon/lib/index.js');

var _nextIcon2 = _interopRequireDefault(_nextIcon);

var _nextProgress = require('../../../next-progress/lib/index.js');

var _nextProgress2 = _interopRequireDefault(_nextProgress);

var _nextUtil = require('../../../next-util/lib/index.js');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/** Step.Item */
var StepItem = (_temp = _class = function (_Component) {
    (0, _inherits3.default)(StepItem, _Component);

    function StepItem(props) {
        (0, _classCallCheck3.default)(this, StepItem);

        var _this = (0, _possibleConstructorReturn3.default)(this, _Component.call(this, props));

        _this.onClick = function () {
            var _this$props = _this.props,
                index = _this$props.index,
                disabled = _this$props.disabled,
                readOnly = _this$props.readOnly,
                animation = _this$props.animation;

            if (disabled || readOnly) {
                return false;
            }

            if (animation && _this.stepNode) {
                _nextUtil.dom.hasClass(_this.stepNode, 'clicked') ? _nextUtil.dom.removeClass(_this.stepNode, 'clicked') : _nextUtil.dom.addClass(_this.stepNode, 'clicked');
            }
            _this.props.onClick(index);
        };

        _this.removeClickedCls = _this.removeClickedCls.bind(_this);
        _this._refHandlerCreator = _this._refHandlerCreator.bind(_this);
        _this.resize = _this.resize.bind(_this);
        return _this;
    }

    StepItem.prototype.componentDidMount = function componentDidMount() {
        var _props = this.props,
            shape = _props.shape,
            direction = _props.direction,
            labelPlacement = _props.labelPlacement,
            index = _props.index,
            total = _props.total;

        if (shape === 'arrow') {
            return;
        }

        if (direction === 'vertical') {
            this.resize();
            _nextUtil.events.on(window, 'resize', this.resize); // 调整垂直Step
        } else if (direction === 'horizontal' && labelPlacement === 'horizontal' && index !== total - 1) {
            // 调整横向Content
            this.adjustTail();
        }
    };

    StepItem.prototype.componentDidUpdate = function componentDidUpdate() {
        var _props2 = this.props,
            shape = _props2.shape,
            direction = _props2.direction,
            labelPlacement = _props2.labelPlacement,
            index = _props2.index,
            total = _props2.total;

        if (shape === 'arrow') {
            return;
        }

        if (direction === 'vertical') {
            this.resize();
        } else if (shape === 'circle' && labelPlacement === 'horizontal' && index !== total - 1) {
            // 调整横向Content
            this.adjustTail();
        } else if (index !== total - 1) {
            _nextUtil.dom.setStyle(this.tail, {
                width: '',
                // eslint-disable-next-line
                'top': ''
            });
        }
    };

    StepItem.prototype.adjustTail = function adjustTail() {
        var width = this.container.offsetWidth + this.title.offsetWidth;
        _nextUtil.dom.setStyle(this.tail, {
            width: 'calc(100% - ' + width + 'px)',
            top: _nextUtil.dom.getStyle(this.container, 'height') / 2 + 'px'
        });
    };

    StepItem.prototype.resize = function resize() {
        var _props3 = this.props,
            index = _props3.index,
            total = _props3.total;

        var stepWidth = _nextUtil.dom.getStyle(this.step, 'width');
        var left = _nextUtil.dom.getStyle(this.body, 'left');
        this.body.style.left = stepWidth + 'px';
        _nextUtil.dom.setStyle(this.body, {
            width: _nextUtil.dom.getStyle(this.step.parentNode.parentNode, 'width') - stepWidth
        });
        _nextUtil.dom.setStyle(this.tail, 'height', _nextUtil.dom.getStyle(this.body, 'height') - _nextUtil.dom.getStyle(this.container, 'height'));
    };

    StepItem.prototype._getNode = function _getNode() {
        var _props4 = this.props,
            prefix = _props4.prefix,
            index = _props4.index,
            status = _props4.status,
            icon = _props4.icon,
            shape = _props4.shape,
            percent = _props4.percent,
            itemRender = _props4.itemRender;

        var nodeElement = icon;
        if (shape === 'dot') {
            nodeElement = icon ? _react2.default.createElement(_nextIcon2.default, { type: icon }) : _react2.default.createElement(
                'div',
                { className: prefix + 'step-item-node-circle' },
                ' '
            );
        } else if (shape === 'circle' && percent) {
            nodeElement = _react2.default.createElement(_nextProgress2.default, { shape: 'circle', percent: percent, className: prefix + 'step-item-progress' });
        } else if (shape === 'circle' && !!itemRender && typeof itemRender === 'function') {
            nodeElement = null; // 如果是需要自定义节点，则不处理，返回空
        } else {
            nodeElement = _react2.default.createElement(
                'div',
                { className: prefix + 'step-item-node-circle' },
                icon ? _react2.default.createElement(_nextIcon2.default, { type: icon }) : this._itemRender(index, status)
            );
        }

        return nodeElement;
    };

    StepItem.prototype.getNode = function getNode(args) {
        var _props5 = this.props,
            direction = _props5.direction,
            prefix = _props5.prefix,
            itemRender = _props5.itemRender,
            index = _props5.index,
            status = _props5.status,
            title = _props5.title,
            content = _props5.content;
        var others = args.others,
            stepCls = args.stepCls,
            overlayCls = args.overlayCls;

        var nodeElement = this._getNode();
        var finalNodeElement = _react2.default.createElement(
            'div',
            { className: prefix + 'step-item-container', ref: this._refHandlerCreator('container') },
            _react2.default.createElement(
                'div',
                { className: prefix + 'step-item-node-placeholder', onClick: this.onClick },
                _react2.default.createElement(
                    'div',
                    {
                        className: prefix + 'step-item-node',
                        ref: this._refHandlerCreator('stepNode'),
                        onTransitionEnd: this.removeClickedCls },
                    nodeElement
                )
            )
        );

        if (!nodeElement) {
            // 需要自定义子节点
            finalNodeElement = _react2.default.createElement(
                'div',
                { className: prefix + 'step-item-container' },
                _react2.default.createElement(
                    'div',
                    { className: prefix + 'step-item-node-placeholder', onClick: this.onClick },
                    itemRender(index, status, title, content)
                )
            );
        }

        return _react2.default.createElement(
            'div',
            (0, _extends3.default)({}, others, { style: this.getStyle(), className: stepCls, ref: this._refHandlerCreator('step') }),
            finalNodeElement,
            _react2.default.createElement(
                'div',
                { className: prefix + 'step-item-body', ref: this._refHandlerCreator('body') },
                _react2.default.createElement(
                    'div',
                    { className: prefix + 'step-item-title', ref: this._refHandlerCreator('title') },
                    title
                ),
                _react2.default.createElement(
                    'div',
                    { className: prefix + 'step-item-content' },
                    content
                )
            ),
            _react2.default.createElement(
                'div',
                { className: prefix + 'step-item-tail', ref: this._refHandlerCreator('tail') },
                _react2.default.createElement(
                    'div',
                    { className: prefix + 'step-item-tail-underlay' },
                    _react2.default.createElement('div', { className: prefix + 'step-item-tail-overlay', style: overlayCls })
                )
            )
        );
    };

    StepItem.prototype.getStyle = function getStyle() {
        var _props6 = this.props,
            parentWidth = _props6.parentWidth,
            parentHeight = _props6.parentHeight,
            direction = _props6.direction,
            total = _props6.total,
            index = _props6.index,
            shape = _props6.shape;

        var width = 'auto';

        if (Number(parentWidth) && Number(parentHeight)) {
            if (!_nextUtil.support.flex && shape === 'arrow') {
                width = Math.floor(parentWidth / total - parentHeight / 2 - parentHeight / 8);
            }
        }
        if (shape !== 'arrow' && direction === 'horizontal') {
            width = total - 1 !== index ? Math.floor(100 / total) + '%' : 'auto';
        }
        return {
            width: width
        };
    };

    StepItem.prototype.removeClickedCls = function removeClickedCls() {
        var _props7 = this.props,
            animation = _props7.animation,
            prefix = _props7.prefix;

        if (animation && this.stepNode && _nextUtil.dom.hasClass(this.stepNode, 'clicked')) {
            _nextUtil.dom.removeClass(this.stepNode, 'clicked');
        }
    };

    // 节点的渲染方法


    StepItem.prototype._itemRender = function _itemRender(index, status) {
        var itemRender = this.props.itemRender;

        if (itemRender) {
            return itemRender(index, status);
        }
        return status === 'finish' ? _react2.default.createElement(_nextIcon2.default, { type: 'select' }) : index + 1;
    };

    StepItem.prototype._refHandlerCreator = function _refHandlerCreator(refName) {
        var self = this;
        return function (ref) {
            self[refName] = ref;
        };
    };

    StepItem.prototype.render = function render() {
        var _classNames;

        var _props8 = this.props,
            prefix = _props8.prefix,
            locale = _props8.locale,
            className = _props8.className,
            status = _props8.status,
            title = _props8.title,
            icon = _props8.icon,
            index = _props8.index,
            total = _props8.total,
            shape = _props8.shape,
            content = _props8.content,
            direction = _props8.direction,
            disabled = _props8.disabled,
            onClick = _props8.onClick,
            readOnly = _props8.readOnly,
            animation = _props8.animation,
            parentHeight = _props8.parentHeight,
            itemRender = _props8.itemRender,
            parentWidth = _props8.parentWidth,
            labelPlacement = _props8.labelPlacement,
            others = (0, _objectWithoutProperties3.default)(_props8, ['prefix', 'locale', 'className', 'status', 'title', 'icon', 'index', 'total', 'shape', 'content', 'direction', 'disabled', 'onClick', 'readOnly', 'animation', 'parentHeight', 'itemRender', 'parentWidth', 'labelPlacement']);


        var stepCls = (0, _classnames2.default)((_classNames = {}, _classNames[prefix + 'step-item'] = true, _classNames[prefix + 'step-item-' + status] = status, _classNames[prefix + 'step-item-first'] = index === 0, _classNames[prefix + 'step-item-last'] = index === total - 1, _classNames[prefix + 'step-item-disabled'] = disabled, _classNames[prefix + 'step-item-read-only'] = readOnly, _classNames[className] = className, _classNames));

        var overlayCls = status === 'finish' ? { width: '100%' } : null;
        var arrowElement = _react2.default.createElement(
            'div',
            (0, _extends3.default)({}, others, { style: this.getStyle(), className: stepCls, onClick: this.onClick }),
            _react2.default.createElement(
                'div',
                { className: prefix + 'step-item-container' },
                _react2.default.createElement(
                    'div',
                    { className: prefix + 'step-item-title' },
                    title
                )
            )
        );
        var otherElement = this.getNode({ others: others, stepCls: stepCls, overlayCls: overlayCls });

        return shape === 'arrow' ? arrowElement : otherElement;
    };

    return StepItem;
}(_react.Component), _class.propTypes = {
    /**
     * 组件的样式品牌前缀
     */
    prefix: _propTypes2.default.string,
    /**
     * 步骤的状态，如不传，会根据外层的 Step 的 current 属性生成，可选值为 `wait`, `process`, `finish`
     */
    status: _propTypes2.default.oneOf(['wait', 'process', 'finish']),
    /**
     * 标题
     */
    title: _propTypes2.default.node,
    /**
     * 百分比
     */
    direction: _propTypes2.default.oneOf(['horizontal', 'vertical']),
    /**
     * 横向布局时的内容排列
     */
    labelPlacement: _propTypes2.default.oneOf(['horizontal', 'vertical']),
    shape: _propTypes2.default.oneOf(['circle', 'arrow', 'dot']),
    /**
     * 图标
     */
    icon: _propTypes2.default.string,
    /**
     * 内容，用于垂直状态下的内容填充
     */
    content: _propTypes2.default.node,
    /**
     * StepItem 的自定义渲染
     * @param {Number} index   节点索引
     * @param {Object} status  节点状态
     * @returns {Node} 节点的渲染结果
     */
    itemRender: _propTypes2.default.func,
    percent: _propTypes2.default.number,
    index: _propTypes2.default.number,
    total: _propTypes2.default.number,
    animation: _propTypes2.default.bool, // 是否开启动效，由父级传入
    /**
     * 是否禁用
     */
    disabled: _propTypes2.default.bool,
    parentWidth: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
    parentHeight: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
    /**
     * 点击步骤时的回调
     * @param {Number} index 节点索引
     */
    onClick: _propTypes2.default.func,
    /**
     * 自定义样式
     */
    className: _propTypes2.default.string
}, _class.defaultProps = {
    shape: 'circle',
    index: 0,
    total: 1,
    onClick: function onClick() {}
}, _temp);
StepItem.displayName = 'StepItem';
exports.default = StepItem;
module.exports = exports['default'];
