// aop-ui/lib/_components/@alife/next-upload/lib/upload.js
'use strict';

exports.__esModule = true;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _class, _temp, _initialiseProps;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _nextUtil = require('../../next-util/lib/index.js');

var _nextIcon = require('../../next-icon/lib/index.js');

var _nextIcon2 = _interopRequireDefault(_nextIcon);

var _nextConfigProvider = require('../../next-config-provider/lib/index.js');

var _nextConfigProvider2 = _interopRequireDefault(_nextConfigProvider);

var _base = require('./base.js');

var _base2 = _interopRequireDefault(_base);

var _index = require('./runtime/index.js');

var _index2 = _interopRequireDefault(_index);

var _html5Uploader = require('./runtime/html5-uploader.js');

var _html5Uploader2 = _interopRequireDefault(_html5Uploader);

var _list = require('./list.js');

var _list2 = _interopRequireDefault(_list);

var _util = require('./util.js');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var noop = _nextUtil.func.noop;

/**
 * Upload
 */
var Upload = (_temp = _class = function (_Base) {
    (0, _inherits3.default)(Upload, _Base);

    function Upload(props) {
        (0, _classCallCheck3.default)(this, Upload);

        var _this = (0, _possibleConstructorReturn3.default)(this, _Base.call(this, props));

        _initialiseProps.call(_this);

        var value = void 0;
        if ('value' in props) {
            value = props.value;
        } else {
            value = props.defaultValue;
        }

        _this.state = {
            value: typeof value === 'undefined' ? [] : value
        };

        _this.uploading = false;
        return _this;
    }

    Upload.prototype.componentWillReceiveProps = function componentWillReceiveProps(nextProps) {
        if ('value' in nextProps && !this.uploading) {
            this.setState({
                value: typeof nextProps.value === 'undefined' ? [] : nextProps.value
            });
        }
    };

    /**
     * 对外暴露API, 添加文件
     * @param files
     */
    Upload.prototype.selectFiles = function selectFiles(files) {
        var filesArr = files.length ? Array.prototype.slice.call(files) : [files];

        this.onSelect(filesArr);
    };

    Upload.prototype.uploadFiles = function uploadFiles(files) {
        var fileList = files.filter(function (file) {
            if (file.state === 'selected') {
                file.state = 'uploading';
                return true;
            }
            return false;
        }).map(function (file) {
            return file.originFileObj;
        });

        this.uploaderRef.startUpload(fileList);
    };

    /**
     * 对外暴露api，控制文件上传
     */


    Upload.prototype.startUpload = function startUpload() {
        this.uploadFiles(this.state.value);
    };

    Upload.prototype.replaceFiles = function replaceFiles(old, current) {
        var targetItem = (0, _util.getFileItem)(old, this.state.value);
        if (!targetItem) {
            return;
        }

        current.uid = old.uid;
        targetItem.originFileObj = current;
    };

    Upload.prototype.isUploading = function isUploading() {
        return this.uploading;
    };

    /**
     * 删除文件
     * @param {File} file
     * @return {void}
     */


    /**
     * 取消上传
     * @param {File} file
     * @return {void}
     */


    Upload.prototype.render = function render() {
        var _classNames, _classNames2;

        var _props = this.props,
            listType = _props.listType,
            prefix = _props.prefix,
            dragable = _props.dragable,
            shape = _props.shape,
            className = _props.className,
            style = _props.style,
            useDataURL = _props.useDataURL,
            disabled = _props.disabled,
            limit = _props.limit,
            closable = _props.closable,
            autoUpload = _props.autoUpload,
            beforeUpload = _props.beforeUpload,
            readonly = _props.readonly,
            onRemove = _props.onRemove,
            onCancel = _props.onCancel,
            onPreview = _props.onPreview,
            list = _props.list,
            extraRender = _props.extraRender,
            others = (0, _objectWithoutProperties3.default)(_props, ['listType', 'prefix', 'dragable', 'shape', 'className', 'style', 'useDataURL', 'disabled', 'limit', 'closable', 'autoUpload', 'beforeUpload', 'readonly', 'onRemove', 'onCancel', 'onPreview', 'list', 'extraRender']);


        var cls = (0, _classnames2.default)((_classNames = {}, _classNames[prefix + 'upload'] = true, _classNames[prefix + 'disabled'] = disabled, _classNames[prefix + 'readonly'] = readonly, _classNames[className] = className, _classNames));

        var innerCls = (0, _classnames2.default)((_classNames2 = {}, _classNames2[prefix + 'upload-inner'] = true, _classNames2[prefix + 'hidden'] = this.state.value.length >= limit, _classNames2));

        var children = this.props.children;
        if (shape === 'card') {
            var _classNames3;

            var cardCls = (0, _classnames2.default)((_classNames3 = {}, _classNames3[prefix + 'upload-card'] = true, _classNames3[prefix + 'disabled'] = disabled, _classNames3));
            children = _react2.default.createElement(
                'div',
                { className: cardCls },
                _react2.default.createElement(_nextIcon2.default, { type: 'add', size: 'large' }),
                _react2.default.createElement(
                    'div',
                    { className: prefix + 'upload-text' },
                    children
                )
            );
        }

        var otherAttributes = _nextUtil.obj.pickAttrsWith(this.props, 'data-');
        return _react2.default.createElement(
            'div',
            (0, _extends3.default)({ className: cls, style: style }, otherAttributes),
            _react2.default.createElement(
                _index2.default,
                (0, _extends3.default)({}, others, {
                    beforeUpload: autoUpload ? beforeUpload : _nextUtil.func.noop,
                    dragable: dragable,
                    disabled: disabled,
                    className: innerCls,
                    onSelect: this.onSelect,
                    onDrop: this.onDrop,
                    onProgress: this.onProgress,
                    onSuccess: this.onSuccess,
                    onError: this.onError,
                    ref: this.saveUploaderRef
                }),
                children
            ),
            children ? _react2.default.createElement('br', null) : null,
            listType || list ? _react2.default.createElement(_list2.default, { useDataURL: useDataURL, uploader: this, listType: listType, value: this.state.value,
                closable: closable, onRemove: onRemove,
                onCancel: onCancel, onPreview: onPreview, extraRender: extraRender }) : null
        );
    };

    return Upload;
}(_base2.default), _class.displayName = 'Upload', _class.propTypes = (0, _extends3.default)({}, _html5Uploader2.default.propTypes, _list2.default.propTypes, {
    /**
     * 样式前缀
     */
    prefix: _propTypes2.default.string.isRequired,
    /**
     * 上传的地址
     */
    action: _propTypes2.default.string,
    /**
     * 文件列表
     */
    value: _propTypes2.default.array,
    /**
     * 默认文件列表
     */
    defaultValue: _propTypes2.default.array,
    /**
     * 上传按钮形状
     */
    shape: _propTypes2.default.oneOf(['card']),
    /**
     * 上传列表的样式
     * @enumdesc 文字, 图文, 卡片
     */
    listType: _propTypes2.default.oneOf(['text', 'image', 'card']),
    list: _propTypes2.default.any,
    /**
     * 上传额外传参
     */
    data: _propTypes2.default.oneOfType([_propTypes2.default.object, _propTypes2.default.func]),
    /**
     * 数据格式化函数，配合自定义 action 使用，参数为服务器的响应数据，详见 [formatter](#formater)
     */
    formatter: _propTypes2.default.func,
    /**
     * 最大文件上传个数
     */
    limit: _propTypes2.default.number,
    /**
     * 设置上传超时,单位ms
     */
    timeout: _propTypes2.default.number,
    /**
     * 可选参数，是否支持拖拽上传，`ie10+` 支持。
     */
    dragable: _propTypes2.default.bool,
    closable: _propTypes2.default.bool,
    useDataURL: _propTypes2.default.bool,
    /**
     * 可选参数，是否禁用上传功能
     */
    disabled: _propTypes2.default.bool,
    /**
     * 选择文件回调
     */
    onSelect: _propTypes2.default.func,
    /**
     * 上传中
     */
    onProgress: _propTypes2.default.func,
    /**
     * 上传文件改变时的状态
     * @param {Object} info 文件事件对象
     */
    onChange: _propTypes2.default.func,
    /**
     * 可选参数，上传成功回调函数，参数为请求下响应信息以及文件
     */
    onSuccess: _propTypes2.default.func,
    /**
     * 移除文件回调函数，详见 [onRemove](#onRemove)
     */
    onRemove: _propTypes2.default.func,
    /**
     * 可选参数，上传失败回调函数，参数为上传失败的信息、响应信息以及文件
     */
    onError: _propTypes2.default.func,
    /**
     * 放文件
     */
    onDrop: _propTypes2.default.func,
    /**
     * 自定义class
     */
    className: _propTypes2.default.string,
    /**
     * 自定义内联样式
     */
    style: _propTypes2.default.object,
    /**
     * 子元素
     */
    children: _propTypes2.default.node
}), _class.defaultProps = (0, _extends3.default)({}, _html5Uploader2.default.defaultProps, _list2.default.defaultProps, {
    prefix: 'next-',
    limit: Infinity,
    autoUpload: true,
    closable: true,
    onSelect: noop,
    onProgress: noop,
    onChange: noop,
    onSuccess: noop,
    onRemove: noop,
    onError: noop,
    onDrop: noop,
    beforeUpload: noop
}), _initialiseProps = function _initialiseProps() {
    var _this2 = this;

    this.onSelect = function (files) {
        var _props2 = _this2.props,
            autoUpload = _props2.autoUpload,
            beforeUpload = _props2.beforeUpload,
            onSelect = _props2.onSelect,
            limit = _props2.limit;


        var fileList = files.map(function (file) {
            var objFile = (0, _util.fileToObject)(file);
            objFile.state = 'selected';
            return objFile;
        });

        var total = _this2.state.value.length + fileList.length;

        if (total > limit) {
            var more = total - limit;
            fileList.splice(fileList.length - more, more);
        }

        var value = _this2.state.value.concat(fileList);

        /* eslint-disable-next */
        _this2.state.value = value;

        if (autoUpload) {
            _this2.uploadFiles(fileList);
        }

        onSelect(fileList, value);

        if (!autoUpload) {
            fileList.forEach(function (file) {
                var before = beforeUpload(file);
                _nextUtil.func.promiseCall(before, _nextUtil.func.noop, function () {
                    _this2.onError(null, null, file); //TODO: handle error message
                });
            });
            _this2.onChange(value, fileList);
        }
    };

    this.onDrop = function (files) {
        _this2.onSelect(files);
    };

    this.onProgress = function (e, file) {
        _this2.uploading = true;

        var value = _this2.state.value;
        var targetItem = (0, _util.getFileItem)(file, value);

        if (!targetItem) {
            return;
        }

        (0, _extends3.default)(targetItem, {
            state: 'uploading',
            percent: e.percent
        });

        _this2.setState({
            value: value
        });

        _this2.props.onProgress(value, targetItem);
    };

    this.onSuccess = function (response, file) {
        _this2.uploading = false;

        var formatter = _this2.props.formatter;


        if (formatter) {
            response = formatter(response);
        }

        try {
            if (typeof response === 'string') {
                response = JSON.parse(response);
            }
        } catch (e) {
            _this2.onError(e, response, file);
            return;
        }

        if (response.success === false) {
            return _this2.onError(response.message, response, file);
        }

        var value = _this2.state.value;
        var targetItem = (0, _util.getFileItem)(file, value);

        if (!targetItem) {
            return;
        }

        (0, _extends3.default)(targetItem, {
            state: 'done',
            response: response,
            url: response.url,
            downloadURL: response.downloadURL || response.url // 下载地址(可选)
        });

        if (!_this2.props.useDataURL) {
            targetItem.imgURL = response.imgURL || response.url; // 缩略图地址(可选)
        }

        _this2.props.onSuccess(targetItem, value);
        _this2.onChange(value, targetItem);
    };

    this.onError = function (err, response, file) {
        _this2.uploading = false;

        var value = _this2.state.value;
        var targetItem = (0, _util.getFileItem)(file, value);

        if (!targetItem) {
            return;
        }

        (0, _extends3.default)(targetItem, {
            state: 'error',
            error: err,
            response: response
        });

        _this2.props.onError(targetItem, value);
        _this2.onChange(value, targetItem);
    };

    this.removeFile = function (file) {
        file.state = 'removed';
        _this2.uploaderRef.abort(file); // 删除组件时调用组件的 `abort` 方法中断上传

        var fileList = _this2.state.value;
        var targetItem = (0, _util.getFileItem)(file, fileList);
        var index = fileList.indexOf(targetItem);
        if (index !== -1) {
            fileList.splice(index, 1);
            _this2.onChange(fileList, targetItem);
        }
    };

    this.abort = function (file) {
        var fileList = _this2.state.value;
        var targetItem = (0, _util.getFileItem)(file, fileList);
        var index = fileList.indexOf(targetItem);
        if (index !== -1) {
            fileList.splice(index, 1);
            _this2.onChange(fileList, targetItem);
        }
        _this2.uploaderRef.abort(file); // 取消上传时调用组件的 `abort` 方法中断上传
    };

    this.onChange = function (value, file) {
        // not controlled
        // if (!('value' in this.props)) {
        //     this.setState({
        //         value
        //     });
        // }
        _this2.setState({
            value: value
        });
        _this2.props.onChange(value, file);
    };
}, _temp);
exports.default = _nextConfigProvider2.default.config(Upload);
module.exports = exports['default'];
