// aop-ui/lib/_components/@alife/next-menu-button/lib/index.js
'use strict';

exports.__esModule = true;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _class, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames3 = require('classnames');

var _classnames4 = _interopRequireDefault(_classnames3);

var _nextButton = require('../../next-button/lib/index.js');

var _nextButton2 = _interopRequireDefault(_nextButton);

var _nextIcon = require('../../next-icon/lib/index.js');

var _nextIcon2 = _interopRequireDefault(_nextIcon);

var _nextMenu = require('../../next-menu/lib/index.js');

var _nextMenu2 = _interopRequireDefault(_nextMenu);

var _nextOverlay = require('../../next-overlay/lib/index.js');

var _nextOverlay2 = _interopRequireDefault(_nextOverlay);

var _nextUtil = require('../../next-util/lib/index.js');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Popup = _nextOverlay2.default.Popup;

/**
 * MenuButton
 */

var MenuButton = (_temp = _class = function (_React$Component) {
    (0, _inherits3.default)(MenuButton, _React$Component);

    function MenuButton(props, context) {
        (0, _classCallCheck3.default)(this, MenuButton);

        var _this = (0, _possibleConstructorReturn3.default)(this, _React$Component.call(this, props, context));

        _this.onPopupOpen = function () {
            var button = (0, _reactDom.findDOMNode)(_this);
            if (_this.props.autoWidth && button && _this.menu) {
                _this.menu.style.width = button.offsetWidth + 'px';
            }
        };

        _this.onPopupVisibleChange = function (visible, type) {
            if (!('visible' in _this.props)) {
                _this.setState({
                    visible: visible
                });
            }
            _this.props.onVisibleChange(visible, type);
        };

        _this._menuRefHandler = function (ref) {
            _this.menu = (0, _reactDom.findDOMNode)(ref);
        };

        _this.state = {
            visible: props.visible || props.defaultVisible
        };
        return _this;
    }

    MenuButton.prototype.componentWillReceiveProps = function componentWillReceiveProps(nextProps) {
        if ('visible' in nextProps) {
            this.setState({
                visible: nextProps.visible
            });
        }
    };

    MenuButton.prototype.render = function render() {
        var _classnames, _classnames2;

        var _props = this.props,
            prefix = _props.prefix,
            style = _props.style,
            className = _props.className,
            label = _props.label,
            popupTriggerType = _props.popupTriggerType,
            popupContainer = _props.popupContainer,
            popupStyle = _props.popupStyle,
            popupClassName = _props.popupClassName,
            popupProps = _props.popupProps,
            selectMode = _props.selectMode,
            onItemClick = _props.onItemClick,
            onSelect = _props.onSelect,
            menuProps = _props.menuProps,
            children = _props.children,
            others = (0, _objectWithoutProperties3.default)(_props, ['prefix', 'style', 'className', 'label', 'popupTriggerType', 'popupContainer', 'popupStyle', 'popupClassName', 'popupProps', 'selectMode', 'onItemClick', 'onSelect', 'menuProps', 'children']);


        var state = this.state;

        var classNames = (0, _classnames4.default)((_classnames = {}, _classnames[prefix + 'menu-btn'] = true, _classnames[prefix + 'expand'] = state.visible, _classnames.opened = state.visible, _classnames), className);

        var popupClassNames = (0, _classnames4.default)((_classnames2 = {}, _classnames2[prefix + 'menu-btn-popup'] = true, _classnames2), popupClassName);

        var trigger = _react2.default.createElement(
            _nextButton2.default,
            (0, _extends3.default)({ style: style, className: classNames }, _nextUtil.obj.pickOthers(MenuButton.propTypes, others)),
            label,
            ' ',
            _react2.default.createElement(_nextIcon2.default, { type: 'arrow-down', className: prefix + 'menu-btn-arrow' })
        );

        return _react2.default.createElement(
            Popup,
            (0, _extends3.default)({}, popupProps, {
                visible: state.visible,
                onVisibleChange: this.onPopupVisibleChange,
                trigger: trigger,
                triggerType: popupTriggerType,
                container: popupContainer,
                onOpen: this.onPopupOpen,
                style: popupStyle,
                className: popupClassNames
            }),
            _react2.default.createElement(
                _nextMenu2.default,
                (0, _extends3.default)({}, menuProps, {
                    ref: this._menuRefHandler,
                    selectMode: selectMode,
                    onSelect: onSelect,
                    onItemClick: onItemClick }),
                children
            )
        );
    };

    return MenuButton;
}(_react2.default.Component), _class.propTypes = {
    prefix: _propTypes2.default.string,
    /**
     * 按钮上的文本内容
     */
    label: _propTypes2.default.node,
    /**
     * 弹层是否与按钮宽度相同
     */
    autoWidth: _propTypes2.default.bool,
    /**
     * 弹层触发方式
     */
    popupTriggerType: _propTypes2.default.oneOf(['click', 'hover']),
    /**
     * 弹层容器
     */
    popupContainer: _propTypes2.default.func,
    /**
     * 弹层展开状态
     */
    visible: _propTypes2.default.bool,
    /**
     * 弹层默认是否展开
     */
    defaultVisible: _propTypes2.default.bool,
    /**
     * 弹层在显示和隐藏触发的事件
     */
    onVisibleChange: _propTypes2.default.func,
    /**
     * 弹层自定义样式
     */
    popupStyle: _propTypes2.default.object,
    /**
     * 弹层自定义样式类
     */
    popupClassName: _propTypes2.default.string,
    /**
     * 弹层属性透传
     */
    popupProps: _propTypes2.default.object,
    /**
     * 菜单的选择模式，同 Menu
     */
    selectMode: _propTypes2.default.string,
    /**
     * 点击菜单项后的回调，同 Menu
     */
    onItemClick: _propTypes2.default.func,
    /**
     * 选择菜单后的回调，同 Menu
     */
    onSelect: _propTypes2.default.func,
    /**
     * 菜单属性透传
     */
    menuProps: _propTypes2.default.object,
    style: _propTypes2.default.object,
    className: _propTypes2.default.string,
    children: _propTypes2.default.any
}, _class.defaultProps = {
    prefix: 'next-',
    autoWidth: true,
    popupTriggerType: 'click',
    onVisibleChange: _nextUtil.func.noop,
    onItemClick: _nextUtil.func.noop,
    onSelect: _nextUtil.func.noop
}, _temp);
MenuButton.displayName = 'MenuButton';


MenuButton.Item = _nextMenu2.default.Item;
MenuButton.Group = _nextMenu2.default.Group;
MenuButton.Divider = _nextMenu2.default.Divider;

exports.default = MenuButton;
module.exports = exports['default'];
