// aop-ui/lib/_components/@alife/next-util/lib/index.js
'use strict';

exports.__esModule = true;
exports.KEYCODE = exports.guid = exports.focus = exports.support = exports.str = exports.obj = exports.log = exports.func = exports.events = exports.env = exports.dom = undefined;

var _dom2 = require('./dom.js');

var _dom = _interopRequireWildcard(_dom2);

var _env2 = require('./env.js');

var _env = _interopRequireWildcard(_env2);

var _events2 = require('./events.js');

var _events = _interopRequireWildcard(_events2);

var _func2 = require('./func.js');

var _func = _interopRequireWildcard(_func2);

var _log2 = require('./log.js');

var _log = _interopRequireWildcard(_log2);

var _object2 = require('./object.js');

var _object = _interopRequireWildcard(_object2);

var _string2 = require('./string.js');

var _string = _interopRequireWildcard(_string2);

var _support2 = require('./support.js');

var _support = _interopRequireWildcard(_support2);

var _focus2 = require('./focus.js');

var _focus = _interopRequireWildcard(_focus2);

var _guid2 = require('./guid.js');

var _guid3 = _interopRequireDefault(_guid2);

var _keycode = require('./keycode.js');

var _keycode2 = _interopRequireDefault(_keycode);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

var dom = exports.dom = _dom; /**
                               * @alife/next-util
                               * @author Fusion Team<http://next.alibaba-inc.com>
                               * @create 2017/09/11
                               * @exports
                               * <Object<Function|*>> dom, env, events, func, log, object, string, focus
                               * <Object<String|Boolean>> support
                               * <Object<String>> KEYCODE
                               * <Function> guid
                               */

var env = exports.env = _env;
var events = exports.events = _events;
var func = exports.func = _func;
var log = exports.log = _log;
var obj = exports.obj = _object;
var str = exports.str = _string;
var support = exports.support = _support;
var focus = exports.focus = _focus;
var guid = exports.guid = _guid3.default;
var KEYCODE = exports.KEYCODE = _keycode2.default;
