// aop-ui/lib/_components/@alife/next-split-button/lib/index.js
'use strict';

exports.__esModule = true;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _class, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames3 = require('classnames');

var _classnames4 = _interopRequireDefault(_classnames3);

var _nextIcon = require('../../next-icon/lib/index.js');

var _nextIcon2 = _interopRequireDefault(_nextIcon);

var _nextButton = require('../../next-button/lib/index.js');

var _nextButton2 = _interopRequireDefault(_nextButton);

var _nextOverlay = require('../../next-overlay/lib/index.js');

var _nextOverlay2 = _interopRequireDefault(_nextOverlay);

var _nextMenu = require('../../next-menu/lib/index.js');

var _nextMenu2 = _interopRequireDefault(_nextMenu);

var _nextConfigProvider = require('../../next-config-provider/lib/index.js');

var _nextConfigProvider2 = _interopRequireDefault(_nextConfigProvider);

var _nextUtil = require('../../next-util/lib/index.js');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Popup = _nextOverlay2.default.Popup;

/**
 * SplitButton
 */

var SplitButton = (_temp = _class = function (_React$Component) {
    (0, _inherits3.default)(SplitButton, _React$Component);

    function SplitButton(props, context) {
        (0, _classCallCheck3.default)(this, SplitButton);

        var _this = (0, _possibleConstructorReturn3.default)(this, _React$Component.call(this, props, context));

        _this.onPopupOpen = function () {
            if (_this.props.autoWidth && _this.wrapper && _this.menu) {
                _nextUtil.dom.setStyle(_this.menu, {
                    width: _this.wrapper.offsetWidth
                });
            }
        };

        _this.onVisibleChange = function (visible, reason) {
            if (!('visible' in _this.props)) {
                _this.setState({
                    visible: visible
                });
            }
            _this.props.onVisibleChange(visible, reason);
        };

        _this._menuRefHandler = function (ref) {
            _this.menu = (0, _reactDom.findDOMNode)(ref);
        };

        _this._wrapperRefHandler = function (ref) {
            _this.wrapper = (0, _reactDom.findDOMNode)(ref);
        };

        _this.state = {
            visible: props.visible || props.defaultVisible
        };
        return _this;
    }

    SplitButton.prototype.componentDidMount = function componentDidMount() {
        // 由于定位目标是 wrapper，如果弹层默认展开，wrapper 还未渲染，didMount 后强制再渲染一次，弹层重新定位
        if (this.state.visible) {
            this.forceUpdate();
        }
    };

    SplitButton.prototype.componentWillReceiveProps = function componentWillReceiveProps(nextProps) {
        if ('visible' in nextProps) {
            this.setState({
                visible: nextProps.visible
            });
        }
    };

    SplitButton.prototype.render = function render() {
        var _classnames,
            _classnames2,
            _this2 = this;

        var _props = this.props,
            prefix = _props.prefix,
            label = _props.label,
            size = _props.size,
            type = _props.type,
            component = _props.component,
            ghost = _props.ghost,
            className = _props.className,
            style = _props.style,
            children = _props.children,
            triggerProps = _props.triggerProps,
            popupTriggerType = _props.popupTriggerType,
            popupStyle = _props.popupStyle,
            popupClassName = _props.popupClassName,
            popupProps = _props.popupProps,
            selectMode = _props.selectMode,
            onSelect = _props.onSelect,
            onItemClick = _props.onItemClick,
            menuProps = _props.menuProps,
            disabled = _props.disabled,
            others = (0, _objectWithoutProperties3.default)(_props, ['prefix', 'label', 'size', 'type', 'component', 'ghost', 'className', 'style', 'children', 'triggerProps', 'popupTriggerType', 'popupStyle', 'popupClassName', 'popupProps', 'selectMode', 'onSelect', 'onItemClick', 'menuProps', 'disabled']);
        var visible = this.state.visible;


        var classNames = (0, _classnames4.default)((_classnames = {}, _classnames[prefix + 'split-btn'] = true, _classnames), className);

        var sharedBtnProps = {
            type: type,
            size: size,
            component: component,
            ghost: ghost,
            disabled: disabled
        };

        var triggerClassNames = (0, _classnames4.default)((_classnames2 = {}, _classnames2[prefix + 'split-btn-trigger'] = true, _classnames2[prefix + 'expand'] = visible, _classnames2.opened = visible, _classnames2));

        var trigger = _react2.default.createElement(
            _nextButton2.default,
            (0, _extends3.default)({}, triggerProps, sharedBtnProps, { className: triggerClassNames }),
            _react2.default.createElement(_nextIcon2.default, { type: 'arrow-down' })
        );

        return _react2.default.createElement(
            _nextButton2.default.Group,
            (0, _extends3.default)({}, _nextUtil.obj.pickOthers(SplitButton.propTypes, others), { className: classNames, style: style, size: size, ref: this._wrapperRefHandler }),
            _react2.default.createElement(
                _nextButton2.default,
                sharedBtnProps,
                label
            ),
            _react2.default.createElement(
                Popup,
                (0, _extends3.default)({}, popupProps, {
                    visible: visible,
                    onVisibleChange: this.onVisibleChange,
                    trigger: trigger,
                    triggerType: popupTriggerType,
                    target: function target() {
                        return _this2.wrapper;
                    },
                    style: popupStyle,
                    shouldUpdatePosition: true,
                    className: popupClassName,
                    onOpen: this.onPopupOpen }),
                _react2.default.createElement(
                    _nextMenu2.default,
                    (0, _extends3.default)({}, menuProps, {
                        selectMode: selectMode,
                        onSelect: onSelect,
                        onItemClick: onItemClick,
                        ref: this._menuRefHandler }),
                    children
                )
            )
        );
    };

    return SplitButton;
}(_react2.default.Component), _class.propTypes = {
    prefix: _propTypes2.default.string,
    style: _propTypes2.default.object,
    /**
     * 按钮的类型
     */
    type: _propTypes2.default.oneOf(['normal', 'primary', 'secondary']),
    /**
     * 按钮组的尺寸
     */
    size: _propTypes2.default.oneOf(['small', 'medium', 'large']),
    /**
     * 主按钮的文案
     */
    label: _propTypes2.default.node,
    /**
     * 设置标签类型
     */
    component: _propTypes2.default.oneOf(['button', 'a']),
    /**
     * 是否为幽灵按钮
     */
    ghost: _propTypes2.default.oneOf(['light', 'dark', false, true]),
    /**
     * 菜单的选择模式
     */
    selectMode: _propTypes2.default.oneOf(['single', 'multiple']),
    /**
     * 选择菜单项时的回调，参考 Menu
     */
    onSelect: _propTypes2.default.func,
    /**
     * 点击菜单项时的回调，参考 Menu
     */
    onItemClick: _propTypes2.default.func,
    /**
     * 触发按钮的属性（支持 Button 的所有属性透传）
     */
    triggerProps: _propTypes2.default.object,
    /**
     * 弹层菜单的宽度是否与按钮组一致
     */
    autoWidth: _propTypes2.default.bool,
    /**
     * 弹层是否显示
     */
    visible: _propTypes2.default.bool,
    /**
     * 弹层默认是否显示
     */
    defaultVisible: _propTypes2.default.bool,
    /**
     * 弹层显示状态变化时的回调函数
     * @param {Boolean} visible 弹层显示状态
     * @param {String} reason 触发弹层显示和隐藏的来源
     */
    onVisibleChange: _propTypes2.default.func,
    /**
     * 弹层的触发方式
     */
    popupTriggerType: _propTypes2.default.oneOf(['click', 'hover']),
    /**
     * 弹层对齐方式
     */
    popupAlign: _propTypes2.default.string,
    /**
     * 弹层自定义样式
     */
    popupStyle: _propTypes2.default.object,
    /**
     * 弹层自定义样式类
     */
    popupClassName: _propTypes2.default.string,
    /**
     * 透传给弹层的属性
     */
    popupProps: _propTypes2.default.object,
    className: _propTypes2.default.string,
    children: _propTypes2.default.any
}, _class.defaultProps = {
    prefix: 'next-',
    type: 'normal',
    size: 'medium',
    autoWidth: true,
    popupTriggerType: 'click',
    popupAlign: 'tr br',
    onVisibleChange: _nextUtil.func.noop
}, _temp);
SplitButton.displayName = 'SplitButton';


SplitButton.Item = _nextMenu2.default.Item;
SplitButton.Divider = _nextMenu2.default.Divider;
SplitButton.Group = _nextMenu2.default.Group;

exports.default = _nextConfigProvider2.default.config(SplitButton);
module.exports = exports['default'];
