// aop-ui/lib/_components/@alife/next-card/lib/index.js
'use strict';

exports.__esModule = true;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _class, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _nextIcon = require('../../next-icon/lib/index.js');

var _nextIcon2 = _interopRequireDefault(_nextIcon);

var _nextButton = require('../../next-button/lib/index.js');

var _nextButton2 = _interopRequireDefault(_nextButton);

var _nextConfigProvider = require('../../next-config-provider/lib/index.js');

var _nextConfigProvider2 = _interopRequireDefault(_nextConfigProvider);

var _zhCn = require('../../next-locale/lib/zh-cn.js');

var _zhCn2 = _interopRequireDefault(_zhCn);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/** Card */
var Card = (_temp = _class = function (_React$Component) {
    (0, _inherits3.default)(Card, _React$Component);

    function Card(props, context) {
        (0, _classCallCheck3.default)(this, Card);

        var _this = (0, _possibleConstructorReturn3.default)(this, _React$Component.call(this, props, context));

        _this.handleToggle = function () {
            _this.setState(function (prevState) {
                return {
                    expand: !prevState.expand
                };
            });
        };

        _this._contentRefHandler = function (ref) {
            _this.content = ref;
        };

        _this.state = {
            needMore: false,
            expand: false,
            contentHeight: 'auto'
        };
        return _this;
    }

    Card.prototype.componentDidMount = function componentDidMount() {
        this._setNeedMore();
        this._setContentHeight();
    };

    Card.prototype.componentDidUpdate = function componentDidUpdate() {
        this._setContentHeight();
    };

    // 是否展示 More 按钮
    Card.prototype._setNeedMore = function _setNeedMore() {
        var contentHeight = this.props.contentHeight;

        var childrenHeight = this._getNodeChildrenHeight(this.content);
        this.setState({
            needMore: contentHeight !== 'auto' && childrenHeight > contentHeight
        });
    };

    // 设置 Body 的高度


    Card.prototype._setContentHeight = function _setContentHeight() {
        if (this.state.expand) {
            var childrenHeight = this._getNodeChildrenHeight(this.content);
            this.content.style.height = childrenHeight + 'px'; // get the real height
        } else {
            this.content.style.height = this.props.contentHeight + 'px';
        }
    };

    Card.prototype._getNodeChildrenHeight = function _getNodeChildrenHeight(node) {
        if (!node) {
            return 0;
        }

        var childrenHeight = 0;
        var contentChildNodes = node.childNodes;
        for (var _iterator = contentChildNodes, _isArray = Array.isArray(_iterator), _i = 0, _iterator = _isArray ? _iterator : _iterator[Symbol.iterator]();;) {
            var _ref;

            if (_isArray) {
                if (_i >= _iterator.length) break;
                _ref = _iterator[_i++];
            } else {
                _i = _iterator.next();
                if (_i.done) break;
                _ref = _i.value;
            }

            var childNode = _ref;

            childrenHeight += childNode.offsetHeight;
        }

        return childrenHeight;
    };

    Card.prototype.render = function render() {
        var _classNames, _classNames2;

        var _props = this.props,
            prefix = _props.prefix,
            className = _props.className,
            title = _props.title,
            subTitle = _props.subTitle,
            extra = _props.extra,
            showTitleBullet = _props.showTitleBullet,
            showHeadDivider = _props.showHeadDivider,
            children = _props.children,
            locale = _props.locale,
            others = (0, _objectWithoutProperties3.default)(_props, ['prefix', 'className', 'title', 'subTitle', 'extra', 'showTitleBullet', 'showHeadDivider', 'children', 'locale']);
        var _state = this.state,
            needMore = _state.needMore,
            expand = _state.expand;


        var cardCls = (0, _classnames2.default)((_classNames = {}, _classNames[prefix + 'card'] = true, _classNames[prefix + 'card-show-divider'] = showHeadDivider, _classNames[prefix + 'card-hide-divider'] = !showHeadDivider, _classNames), className);

        var headCls = (0, _classnames2.default)((_classNames2 = {}, _classNames2[prefix + 'card-head'] = true, _classNames2[prefix + 'card-head-show-bullet'] = showTitleBullet, _classNames2));

        delete others.contentHeight;

        var headExtra = extra ? _react2.default.createElement(
            'div',
            { className: prefix + 'card-extra' },
            extra
        ) : null;

        return _react2.default.createElement(
            'div',
            (0, _extends3.default)({}, others, { className: cardCls }),
            title ? _react2.default.createElement(
                'div',
                { className: headCls },
                _react2.default.createElement(
                    'div',
                    { className: prefix + 'card-head-main' },
                    _react2.default.createElement(
                        'div',
                        { className: prefix + 'card-title' },
                        title,
                        subTitle ? _react2.default.createElement(
                            'span',
                            { className: prefix + 'card-subtitle' },
                            subTitle
                        ) : null
                    ),
                    headExtra
                )
            ) : null,
            _react2.default.createElement(
                'div',
                { className: prefix + 'card-body' },
                _react2.default.createElement(
                    'div',
                    { className: prefix + 'card-content', ref: this._contentRefHandler },
                    children
                ),
                needMore ? _react2.default.createElement(
                    'div',
                    { className: prefix + 'card-footer', onClick: this.handleToggle },
                    _react2.default.createElement(
                        _nextButton2.default,
                        { text: true, type: 'primary' },
                        expand ? locale.fold : locale.expand,
                        ' ',
                        _react2.default.createElement(_nextIcon2.default, { type: 'arrow-down', size: 'xs', className: expand ? 'expand' : '' })
                    )
                ) : null
            )
        );
    };

    return Card;
}(_react2.default.Component), _class.displayName = 'Card', _class.propTypes = {
    prefix: _propTypes2.default.string,
    /**
     * 卡片的标题
     */
    title: _propTypes2.default.string,
    /**
     * 卡片的副标题
     */
    subTitle: _propTypes2.default.string,
    /**
     * 是否显示标题的项目符号
     */
    showTitleBullet: _propTypes2.default.bool,
    /**
     * 是否展示头部的分隔线
     */
    showHeadDivider: _propTypes2.default.bool,
    /**
     * 内容区域的固定高度
     */
    contentHeight: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
    /**
     * 标题区域的用户自定义内容
     */
    extra: _propTypes2.default.node,
    className: _propTypes2.default.string
}, _class.defaultProps = {
    prefix: 'next-',
    showTitleBullet: true,
    showHeadDivider: true,
    contentHeight: 120,
    locale: _zhCn2.default.Card
}, _temp);
Card.displayName = 'Card';
exports.default = _nextConfigProvider2.default.config(Card);
module.exports = exports['default'];
