// aop-ui/lib/_components/@alife/next-timeline/lib/view/timeline.js
'use strict';

exports.__esModule = true;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _class, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _nextUtil = require('../../../next-util/lib/index.js');

var _nextConfigProvider = require('../../../next-config-provider/lib/index.js');

var _nextConfigProvider2 = _interopRequireDefault(_nextConfigProvider);

var _zhCn = require('../../../next-locale/lib/zh-cn.js');

var _zhCn2 = _interopRequireDefault(_zhCn);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/** Timeline */
var Timeline = (_temp = _class = function (_Component) {
    (0, _inherits3.default)(Timeline, _Component);

    function Timeline(props, context) {
        (0, _classCallCheck3.default)(this, Timeline);

        var _this = (0, _possibleConstructorReturn3.default)(this, _Component.call(this, props, context));

        _this.state = {
            fold: props.fold
        };
        return _this;
    }

    Timeline.prototype.componentWillReceiveProps = function componentWillReceiveProps(newProps) {
        typeof newProps.fold !== 'undefined' && this.setState({ fold: newProps.fold });
    };

    Timeline.prototype.toggleFold = function toggleFold(folderIndex, total) {
        var fold = this.state.fold;


        if (folderIndex) {
            for (var i = 0; i < fold.length; i++) {
                var _fold$i = fold[i],
                    foldArea = _fold$i.foldArea,
                    foldShow = _fold$i.foldShow;


                if (foldArea[1] && folderIndex === foldArea[1] || !foldArea[1] && folderIndex === total - 1) {
                    fold[i].foldShow = !foldShow;
                }
            }

            this.setState({ fold: fold });
        }
    };

    Timeline.prototype.render = function render() {
        var _this2 = this,
            _classNames;

        var _props = this.props,
            prefix = _props.prefix,
            className = _props.className,
            children = _props.children,
            locale = _props.locale,
            animation = _props.animation,
            others = (0, _objectWithoutProperties3.default)(_props, ['prefix', 'className', 'children', 'locale', 'animation']);
        var fold = this.state.fold;

        // 修改子节点属性

        var childrenCount = _react2.default.Children.count(children);
        var cloneChildren = _react.Children.map(children, function (child, i) {
            var folderIndex = null;
            var foldNodeShow = false;

            fold.forEach(function (item) {
                var foldArea = item.foldArea,
                    foldShow = item.foldShow;


                if (foldArea[0] && i >= foldArea[0] && (i <= foldArea[1] || !foldArea[1])) {
                    folderIndex = foldArea[1] || childrenCount - 1;
                    foldNodeShow = foldShow;
                }
            });

            return _react2.default.cloneElement(child, {
                prefix: prefix,
                locale: locale,
                total: childrenCount,
                index: i,
                folderIndex: folderIndex,
                foldShow: foldNodeShow,
                toggleFold: folderIndex === i ? _this2.toggleFold.bind(_this2, folderIndex, childrenCount) : function () {},
                animation: animation
            });
        });

        var timelineCls = (0, _classnames2.default)((_classNames = {}, _classNames[prefix + 'timeline'] = true, _classNames), className);

        return _react2.default.createElement(
            'ul',
            (0, _extends3.default)({}, _nextUtil.obj.pickOthers(Timeline.propTypes, others), { className: timelineCls }),
            cloneChildren
        );
    };

    return Timeline;
}(_react.Component), _class.propTypes = {
    /**
     * 样式的品牌前缀
     */
    prefix: _propTypes2.default.string,
    /**
     * 自定义折叠选项 示例`[{foldArea: [startIndex, endIndex], foldShow: boolean}]`
     */
    fold: _propTypes2.default.array,
    /**
     * 自定义类名
     */
    className: _propTypes2.default.string,
    children: _propTypes2.default.any,
    locale: _propTypes2.default.object,
    animation: _propTypes2.default.bool
}, _class.defaultProps = {
    prefix: 'next-',
    fold: [],
    locale: _zhCn2.default.Timeline,
    animation: true
}, _temp);
Timeline.displayName = 'Timeline';
exports.default = _nextConfigProvider2.default.config(Timeline);
module.exports = exports['default'];
